scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +159,312 @@
+    (initialize-cluster cluster (:pool-name->fenzo @scheduler-promise))))
+
+(defn execute-update!
+  "Attempt to execute a valid cluster configuration update.
+  Change cluster state if there is an existing cluster in memory.
+  Create clusters if there is an active update an no corresponding cluster in memory. This can happen at startup or
+  when a brand new cluster is added.
+  Reflect changes in the database."
+  [conn
+   {:keys [valid? changed? active?] {:keys [name state state-locked?] :as goal-config} :goal-config}
+   current-in-mem-configs]
+  {:pre [valid?]}
+  (try
+    (let [{:keys [state-atom state-locked?-atom] :as cluster} (@cluster-name->compute-cluster-atom name)
+          current-in-mem-config (current-in-mem-configs name)]
+      ; TODO: keep these checks? check some other way?
+      (when (and current-in-mem-config (not (and cluster state-atom state-locked?-atom)))
+        (throw (ex-info "We know an in-memory config but we don't know the corresponding in-memory cluster.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (when (and (or cluster state-atom state-locked?-atom) (not current-in-mem-config))
+        (throw (ex-info "We know an in-memory cluster but we don't know the corresponding in-memory config.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (if current-in-mem-config
+        (do
+          (reset! state-atom state)
+          (reset! state-locked?-atom state-locked?))
+        (when active? (initialize-cluster! goal-config))))
+    (when changed?
+      ; :db.unique/identity gives upsert behavior (update on insert), so we don't need to specify an entity ID when updating
+      @(d/transact conn [(assoc (compute-cluster-config->compute-cluster-config-ent goal-config)
+                           :db/id (d/tempid :db.part/user))]))
+    {:update-succeeded true}
+    (catch Throwable t
+      (log/error t "Failed to update cluster" goal-config)
+      {:update-succeeded false :error-message (.toString t)})))
+
+(defn update-compute-clusters
+  "This function allows adding or updating the current compute cluster configurations. Takes
+  in a map of configurations with cluster name as the key. These configurations represent the only known
+  configurations, and clusters that are missing from this map are set to the 'deleted' state.
+  Only the state of an existing cluster and its configuration can be changed unless force? is set to true."
+  ([conn new-configs force?]
+   (update-compute-clusters conn nil new-configs force?))
+  ([conn new-config new-configs force?]
+   {:pre [(= 1 (->> [new-config new-configs] (filter some?) count))]}
```

Comment:
So you can only specify one of new-config or new-configs?

This screams two different functions, or maybe one 'update-single-cluster' and 'replace-all-clusters', where the first function calls adds the one new cluster to the existing clusters and then calls the second.

Reply:
I'll do "helper"